odxtools/udsbinner.py

```diff
@@ -0,0 +1,123 @@
+from .parameters.codedconstparameter import CodedConstParameter
+
+
+class UdsBinner:
+    """Class to categorize a long list of services into the service
+    groups defined by the UDS standard.
+
+    This class is supposed to be used like this:
+
+    db = odxtools.load_file("my_cool_diagnostics_db.pdx")
+    ...
+    uds_bins = UdsBinner(db.ecus.my_ecu)
+    uds_bins.print_info()
+    """
+
+    def __init__(self, services: Iterable[DiagService]):
+        service_groups: Dict[Optional[int], NamedItemList[DiagService]] = {}
+        for service in services:
+            SID = self.__extract_sid(service)
+
+            if SID not in service_groups:
+                service_groups[SID] = NamedItemList()
+
+            service_groups[SID].append(service)
+
+        self.service_groups = service_groups
+
+    def __extract_sid(self, service: DiagService) -> Optional[int]:
+        # diagnostic services without requests are possible; just like
+        # aircraft without wings...
+        if service.request is None:
+            return None
+
+        prefix = 0  # prefix of constant parameters
+        cursor = 0  # bit position of the next parameter
+        for param in service.request.parameters:
+            if not isinstance(param, CodedConstParameter):
+                # we *need* at least the first byte of a request to be statically defined!
+                return None
+
+            param_len = param.get_static_bit_length()
+            if param_len is None:
+                return None
+
+            if not isinstance(param.coded_value, int):
+                return None
+
+            prefix <<= param_len
+            prefix |= param.coded_value & ((1 << param_len) - 1)
+            cursor += param_len
+
+            if cursor >= 8:
+                # we have a prefix that is at least 8 bits
+                # long. return its most significant byte.
+                prefix >>= cursor - 8
+                return prefix & 0xff
+
+        return None
+
+    def print_info(self) -> None:
+        """Convenience method to print which of the diagnostic
+        services are part of which UDS service group.
+        """
+        for sid, service_list in self.service_groups.items():
+            sid_name = self.sid_to_name.get(sid)
+            if isinstance(sid, int):
+                if sid_name is not None:
+                    print(f"UDS service group '{sid_name}' (0x{sid:x}):")
+                else:
+                    print(f"Non-standard UDS service group 0x{sid:x}:")
+            else:
+                print(f"Non standard services:")
+            for service in service_list:
+                print(f"  {service.short_name}")
+
+    #: map from the ID of a service group to its human-readable name
+    #: as defined by the UDS standard
+    sid_to_name = {
+        # diagnostic services that cannot be categorized in terms of
+        # the UDS standard get `None` as their SID...
+        None: "Not Applicable",
```

Comment:
```suggestion
        None: "Unknown",
```

Reply:
IMO, in this case, "not applicable" is -- err -- more applicable: This is meant for all services which do not conform to the UDS specification at all, i.e., the first byte of their request cannot be statically determined. "Unknown" SIDs are SIDs which have not (yet?) been specified by UDS (or ones which I missed to add...)

Reply:
- there are other standard diag services other than UDS, for example OBD II (range 1 to 0xa)
- assigning all services that you don't recognize the same name is not nice, instead return sid as hex

Reply:
I refactored the code a bit. I hope these issues are solved with that...